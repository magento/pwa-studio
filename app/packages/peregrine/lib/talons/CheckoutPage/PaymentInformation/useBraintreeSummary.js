import { useQuery } from '@apollo/client';

import { useCartContext } from '../../../context/cart';

import mergeOperations from '../../../util/shallowMerge';
import DEFAULT_OPERATIONS from './braintreeSummary.gql';

const mapBillingAddressData = rawBillingAddressData => {
    if (rawBillingAddressData) {
        const { street, country, region } = rawBillingAddressData;

        return {
            ...rawBillingAddressData,
            street1: street[0],
            street2: street[1],
            country: country.code,
            state: region.label
        };
    } else {
        return {};
    }
};

/**
 * Talon for the braintree summary view.
 *
 * @returns {
 *   billingAddress: {
 *      firstName: String,
 *      lastName: String,
 *      country: String,
 *      street1: String,
 *      street2: String,
 *      city: String,
 *      state: String,
 *      postalCode: String,
 *   },
 *   paymentNonce: {
 *      type: String,
 *      description: String,
 *      details: {
 *          cardType: String,
 *          lastFour: String,
 *          lastTwo: String
 *      },
 *   },
 *   isBillingAddressSame: Boolean,
 *   isLoading: Boolean,
 * }
 */
export const useBraintreeSummary = () => {
    const operations = mergeOperations(DEFAULT_OPERATIONS);

    const { getSummaryDataQuery } = operations;

    const [{ cartId }] = useCartContext();
    const { data: summaryData, loading: isLoading } = useQuery(getSummaryDataQuery, {
        skip: !cartId,
        variables: { cartId }
    });

    const billingAddress = summaryData ? mapBillingAddressData(summaryData.cart.billingAddress) : {};

    const isBillingAddressSame = summaryData ? summaryData.cart.isBillingAddressSame : true;

    const paymentNonce = summaryData ? summaryData.cart.paymentNonce : null;

    return {
        billingAddress,
        isBillingAddressSame,
        isLoading,
        paymentNonce
    };
};
