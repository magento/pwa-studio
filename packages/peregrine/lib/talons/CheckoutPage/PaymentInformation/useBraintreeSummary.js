import { useQuery } from '@apollo/client';

import { useCartContext } from '../../../context/cart';

import mergeOperations from '../../../util/shallowMerge';
import defaultOperations from './braintreeSummary.gql';

const mapBillingAddressData = rawBillingAddressData => {
    if (rawBillingAddressData) {
        const { street, country, region } = rawBillingAddressData;

        return {
            ...rawBillingAddressData,
            street1: street[0],
            street2: street[1],
            country: country.code,
            state: region.code
        };
    } else {
        return {};
    }
};

/**
 * Talon for the braintree summary view.
 *
 * @param {DocumentNode} props.operations operations used by this summary component
 *
 * @returns {
 *   billingAddress: {
 *      firstName: String,
 *      lastName: String,
 *      country: String,
 *      street1: String,
 *      street2: String,
 *      city: String,
 *      state: String,
 *      postalCode: String,
 *   },
 *   paymentNonce: {
 *      type: String,
 *      description: String,
 *      details: {
 *          cardType: String,
 *          lastFour: String,
 *          lastTwo: String
 *      },
 *   },
 *   isBillingAddressSame: Boolean,
 *   isLoading: Boolean,
 * }
 */
export const useBraintreeSummary = (props = {}) => {
    const operations = mergeOperations(defaultOperations, props.operations);
    const { getBraintreeSummaryData } = operations.queries;

    const [{ cartId }] = useCartContext();
    const { data: summaryData, loading: isLoading } = useQuery(
        getBraintreeSummaryData,
        {
            skip: !cartId,
            variables: { cartId }
        }
    );

    const billingAddress = summaryData
        ? mapBillingAddressData(summaryData.cart.billingAddress)
        : {};

    const isBillingAddressSame = summaryData
        ? summaryData.cart.isBillingAddressSame
        : true;

    const paymentNonce = summaryData ? summaryData.cart.paymentNonce : null;

    return {
        billingAddress,
        isBillingAddressSame,
        isLoading,
        paymentNonce
    };
};
