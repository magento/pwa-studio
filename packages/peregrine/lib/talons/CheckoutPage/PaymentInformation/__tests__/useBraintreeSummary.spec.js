import React from 'react';

import { useQuery } from '@apollo/client';

import createTestInstance from '../../../../util/createTestInstance';
import { useBraintreeSummary } from '../useBraintreeSummary';

jest.mock('../../../../context/cart', () => ({
    useCartContext: jest.fn().mockReturnValue([{ cartId: '123' }])
}));

jest.mock('@apollo/client', () => ({
    ...jest.requireActual('@apollo/client'),
    useApolloClient: jest.fn().mockReturnValue({
        cache: {
            policies: {
                addTypePolicies: jest.fn()
            }
        }
    }),
    useQuery: jest.fn().mockReturnValueOnce({
        loading: false,
        data: {
            cart: {
                billingAddress: {
                    firstName: 'Goosey',
                    lastName: 'Goose',
                    country: { code: 'United States of Gooseland' },
                    street: ['12345 Gooseey Blvd', 'Apt 123'],
                    city: { code: 'Goostin' },
                    region: 'Gooseyork',
                    postalCode: '12345',
                    phoneNumber: '1234567890'
                },
                isBillingAddressSame: false,
                paymentNonce: { code: 'braintree', title: 'Braintree' }
            }
        }
    })
}));

const Component = props => {
    const talonProps = useBraintreeSummary(props);

    return <i talonProps={talonProps} />;
};

test('Should return correct shape', () => {
    const tree = createTestInstance(<Component queries={{}} />);
    const { root } = tree;
    const { talonProps } = root.findByType('i').props;

    expect(talonProps).toMatchSnapshot();
});

test('mapBillingAddressData() should return an empty object if the raw billing address data is empty', () => {
    useQuery.mockReturnValueOnce({
        loading: false,
        data: {
            cart: {
                isBillingAddressSame: false,
                paymentNonce: { code: 'braintree', title: 'Braintree' }
            }
        }
    });

    const tree = createTestInstance(<Component queries={{}} />);
    const { root } = tree;
    const { talonProps } = root.findByType('i').props;

    expect(talonProps.billingAddress).toEqual({});
});

test('handles empty summary data returned from query', () => {
    useQuery.mockReturnValueOnce({
        loading: false
    });

    const tree = createTestInstance(<Component queries={{}} />);
    const { root } = tree;
    const { talonProps } = root.findByType('i').props;

    expect(talonProps.billingAddress).toEqual({});
    expect(talonProps.isBillingAddressSame).toBeTruthy();
    expect(talonProps.isLoading).toBeFalsy();
    expect(talonProps.paymentNonce).toBeNull();
});
