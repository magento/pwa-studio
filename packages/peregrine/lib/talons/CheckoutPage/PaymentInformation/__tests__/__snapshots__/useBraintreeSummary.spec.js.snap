// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should return correct shape 1`] = `
Object {
  "billingAddress": Object {
    "city": Object {
      "code": "Goostin",
    },
    "country": "United States of Gooseland",
    "firstName": "Goosey",
    "lastName": "Goose",
    "phoneNumber": "1234567890",
    "postalCode": "12345",
    "region": "Gooseyork",
    "state": undefined,
    "street": Array [
      "12345 Gooseey Blvd",
      "Apt 123",
    ],
    "street1": "12345 Gooseey Blvd",
    "street2": "Apt 123",
  },
  "isBillingAddressSame": false,
  "isLoading": false,
  "paymentNonce": Object {
    "code": "braintree",
    "title": "Braintree",
  },
}
`;
